src/lib/statistics/cardinality_estimator.cpp

```diff
@@ -782,6 +785,90 @@ std::shared_ptr<TableStatistics> CardinalityEstimator::estimate_inner_equi_join(
   return output_table_statistics;
 }
 
+std::shared_ptr<TableStatistics> CardinalityEstimator::estimate_semi_join(
+    const ColumnID left_column_id, const ColumnID right_column_id, const TableStatistics& left_input_table_statistics,
+    const TableStatistics& right_input_table_statistics) {
+  // This is based on estimate_inner_equi_join. We take the histogram from the right, set the bin heights to the
+  // distinct counts and run an inner/equi estimation on it. As there are no more duplicates on the right side, we
+  // should get the correct estimation for the left side.
+  const auto left_data_type = left_input_table_statistics.column_data_type(left_column_id);
+  const auto right_data_type = right_input_table_statistics.column_data_type(right_column_id);
+
+  // TODO(anybody) - Implement join estimation for differing column data types
+  //               - Implement join estimation for String columns
+  if (left_data_type != right_data_type || left_data_type == DataType::String) {
+    return estimate_cross_join(left_input_table_statistics, right_input_table_statistics);
+  }
+
+  std::shared_ptr<TableStatistics> output_table_statistics;
+
+  resolve_data_type(left_data_type, [&](const auto data_type_t) {
+    using ColumnDataType = typename decltype(data_type_t)::type;
+
+    const auto left_input_column_statistics = std::dynamic_pointer_cast<AttributeStatistics<ColumnDataType>>(
+        left_input_table_statistics.column_statistics[left_column_id]);
+    const auto right_input_column_statistics = std::dynamic_pointer_cast<AttributeStatistics<ColumnDataType>>(
+        right_input_table_statistics.column_statistics[right_column_id]);
+
+    auto cardinality = Cardinality{0};
+    auto join_column_histogram = std::shared_ptr<AbstractHistogram<ColumnDataType>>{};
+
+    auto left_histogram = left_input_column_statistics->histogram;
+    auto right_histogram = right_input_column_statistics->histogram;
+
+    if (left_histogram && right_histogram) {
+      // Adapt the right histogram so that it only covers distinct values (i.e., replacing the bins' height with their
+      // number of distinct counts)
+      auto distinct_right_histogram_builder =
+          GenericHistogramBuilder(right_histogram->bin_count(), right_histogram->domain());
+      const auto right_bin_count = right_histogram->bin_count();
+      for (auto bin_id = BinID{0}; bin_id < right_bin_count; ++bin_id) {
+        const auto& right_bin = right_histogram->bin(bin_id);
+        distinct_right_histogram_builder.add_bin(right_bin.min, right_bin.max, right_bin.distinct_count,
+                                                 right_bin.distinct_count);
+      }
+
+      const auto distinct_right_histogram = distinct_right_histogram_builder.build();
+      join_column_histogram = estimate_inner_equi_join_with_histograms(*left_histogram, *distinct_right_histogram);
```

Comment:
I was looking at `estimate_inner_equi_join_with_histograms()` and there appears to be a lot of duplicated code. Can we reduce this?

Reply:
You mean between estimate_inner_equi_join and estimate_semi_join? Yes, there is, but I worry that deduplicating this will make it even harder to read.